### PR TITLE
[CloudMigrations] Fetch cloud migration status asynchronously

### PR DESCRIPTION
**Pull request by MissingRoberto**
**What is this feature?**

Fetch cloud migration status asynchronously to the API call. 

**Why do we need this feature?**

The snapshot/ endpoint to get snapshot status during upload and migration becomes very slow (more than 30 seconds) and times out in the k6 script. From the user perspective there is no visible impact if they don't refresh the page, but if they refresh the page , they won't be able to see the state of the snapshot for a very long time or until it finishes.

This is due to the synchronous fetching results of the migration on that endpoint:

https://github.com/grafana/grafana/blob/b0e6539f81d79295636f1774f795484cc461fc9b/pkg/services/cloudmigration/cloudmigrationimpl/cloudmigration.go?plain=1#L549-L591

That synchronous way of fetching the information can also result in unintended consequences as you may have 2 or more requests to that endpoint that trigger that fetching of information and as far as I understood we don't have any locking mechanism.

More context can be found in: https://github.com/grafana/grafana-operator-experience-squad/issues/1032#issuecomment-2459803194

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana-operator-experience-squad/issues/1098

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

**Review by macabu — APPROVED** [review]

File: pkg/services/cloudmigration/cloudmigrationimpl/cloudmigration.go [hunk 1]
```diff
@@ -595,11 +598,74 @@ func (s *Service) GetSnapshot(ctx context.Context, query cloudmigration.GetSnaps
...
+		s.log.Info("synchronize snapshot status already running", "sessionUID", session.UID, "snapshotUID", snapshot.UID)
+		return
+	}
+	defer atomic.StoreInt32(&s.isSyncSnapshotStatusFromGMSRunning, 0)
+
+	if !snapshot.ShouldQueryGMS() {
+		return
+	}
+
+	s.cancelMutex.Lock()
```
[the remark this answers is not in the document]
Comment (MissingRoberto) [reply pkg/services/cloudmigration/cloudmigrationimpl/cloudmigration.go:635]: Regarding naming and structure, the existing code for cancellation appears to be designed not just for canceling uploads, but for canceling any in-progress operation for a given snapshot. For instance, it's used before the upload during the building process.

https://github.com/grafana/grafana/blob/a16db7e1130114920e824f802cd8de51c7bb6010/pkg/services/cloudmigration/cloudmigrationimpl/cloudmigration.go?plain=1#L506-L507

You're correct that this code currently serves no purpose, as the cancel button isn't displayed in the UI once the snapshot is uploaded.

The new lines currently serve two purposes:

1. They're in place for potential future implementation of cancellation for this step.
2. They allow us to stop the goroutine in tests.

To address the second point, I see these alternatives:

1. Use the existing `cancelFunc` logic.
2. Add an explicit `Shutdown` or `Stop` method in the service to stop any goroutine.
3. Pass application context in the service initialization and use `WithCancel` in the test to stop it.

I'm leaning towards option 1, as it introduces fewer changes and aligns with existing code, or option 3, as it's more idiomatic and cleaner.

Which option do you prefer?

File: pkg/services/cloudmigration/cloudmigrationimpl/cloudmigration.go [hunk 2]
```diff
@@ -53,6 +54,8 @@ type Service struct {
 	cancelMutex sync.Mutex
 	cancelFunc  context.CancelFunc
 
+	isSyncSnapshotStatusFromGMSRunning int32
```
[the remark this answers is not in the document]
Comment (MissingRoberto) [reply pkg/services/cloudmigration/cloudmigrationimpl/cloudmigration.go:57]: No reason, let me change it :)

**Comment by MissingRoberto** [comment]
@macabu as I have already the existing code, let's merge it as it is but continue the discussion so that they can be included in the follow-up :)

File: pkg/services/cloudmigration/cloudmigrationimpl/cloudmigration.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (macabu) [reply pkg/services/cloudmigration/cloudmigrationimpl/cloudmigration.go:635]: We can keep it as-is (use option 1), I just wanted to understand its usage 👍

**Pull request merged**